knimin/lib/data_access.py

```diff
@@ -2716,17 +2701,292 @@ def delete_samples(self, ids):
                                      % sample_id)
-                sql = """DELETE FROM pm.sample WHERE sample_id = %s
+                sql = """SELECT sample_plate_id
+                         FROM pm.sample_plate_layout
+                         WHERE sample_id = %s"""
+                TRN.add(sql, [sample_id])
+                res = TRN.execute_fetchindex()
+                if res:
+                    sample_plate_ids = ', '.join([str(x[0]) for x in res])
+                    raise ValueError('Sample ID %s cannot be deleted because '
+                                     'it is associated with sample plate(s) %s'
+                                     '.' % (sample_id, sample_plate_ids))
+                sql = """DELETE FROM pm.sample
+                         WHERE sample_id = %s
                          RETURNING sample_id"""
                 TRN.add(sql, [sample_id])
 
+    def _sample_plate_exists(self, id):
+        """Confirms that a sample plate ID exists
+
+        Parameters
+        ----------
+        id : int
+
+        Raises
+        ------
+        ValueError
+            If the sample plate ID does not exist
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.sample_plate
+                                WHERE sample_plate_id = %s)"""
+        if not self._con.execute_fetchone(sql, [id])[0]:
+            raise ValueError('Sample plate ID %s does not exist.' % id)
+
+    def _sample_plate_is_unique(self, name, skip_id=None):
+        """Confirms that a sample plate name is not duplicate
+
+        Parameters
+        ----------
+        name : str
+        skip_id : int
+            Skip this ID in searching
+
+        Raises
+        ------
+        ValueError
+            If the sample plate name already exists
+        """
+        sql = """SELECT sample_plate_id
+                 FROM pm.sample_plate
+                 WHERE name = %s
+                 AND sample_plate_id IS DISTINCT FROM %s"""
+        res = self._con.execute_fetchone(sql, [name, skip_id])
+        if res:
+            raise ValueError('Name %s conflicts with exisiting sample plate '
+                             '%s.' % (repr(name), res[0]))
+
+    def _email_exists(self, email):
+        """Confirms that an email exists
+
+        Parameters
+        ----------
+        barcode : str
+
+        Raises
+        ------
+        ValueError
+            If the email does not exist
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM ag.labadmin_users
+                                WHERE email = %s)"""
+        if not self._con.execute_fetchone(sql, [email])[0]:
+            raise ValueError('Email %s does not exist.' % email)
+
+    def _plate_type_name_to_id(self, name):
+        """Gets plate type ID by name
+
+        Parameters
+        ----------
+        name : str
+            Name of the plate type. If None, ID of the first plate type (i.e.,
+            96-well) will return
```

Comment:
This is when the user clicks the "Create New Plate" button. I think instead of popping a dialog box asking them which plate type they like, the user may prefer to be directly navigated to a plate map with the default plate type (96-well). This saves one or two user actions. What do you think?